CORDA-3597 Replace old Checkpoint table with new one.

williamvigorr3 opened the pull request.

williamvigorr3 committed:
Replace old Checkpoint table with new one.

Adds some of the new fields into the table where needed (I have
guessed this stuff but we can update it as we go along).

williamvigorr3 committed:
Fix database constraints + name table correctly opps.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -183,58 +186,56 @@ class DBCheckpointStorage : CheckpointStorage {
-    ) {
-        override fun toString() = "DBCheckpoint(checkpointId = ${checkpointId}, checkpointSize = ${checkpoint.size})"
+    private fun createDBCheckpoint(id: StateMachineRunId, checkpoint: Checkpoint,
+                                   serializedCheckpoint: SerializedBytes<Checkpoint>): DBFlowCheckpoint {
```
Maybe pass in the serialization context instead of the serialized bytes. This can be a lazy half way house instead of fully tidying up the dependency injection to this class. A.k.a we need the serialization context when creating `DBCheckpointStorage` to fully fix this.

lankydan replied:
Ill create a jira for tidying this up

williamvigorr3 replied:
I have done this. Annoyingly there was some metering stuff which means we are now serializing the checkpoint twice (one to measure the size). Do you think we should pull the metering stuff into DBCheckpointStorage (that feels more natural to me)?

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -47,27 +47,27 @@ class DBCheckpointStorage : CheckpointStorage {
-            @JoinColumn(name = "flow_id", referencedColumnName = "flow_id")
-            var flowMetadata: DBFlowMetadata,
+//            @OneToOne(fetch = FetchType.LAZY)
+//            @JoinColumn(name = "flow_id", referencedColumnName = "flow_id")
```
Hopefully we can re-add and use `session.get`

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -81,16 +81,17 @@ class DBCheckpointStorage : CheckpointStorage {
+            @GeneratedValue(strategy = GenerationType.SEQUENCE)
             @Column(name = "id", nullable = false)
-            var id: BigInteger? = null,
+            private var id: Long? = null,
```
Can these id's be set to `0` instead of `null` to generate the sequence number?

williamvigorr3 replied:
Yes I have got this working now.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -47,27 +47,27 @@ class DBCheckpointStorage : CheckpointStorage {
             var blob: DBFlowCheckpointBlob,
 
-            @OneToOne(fetch = FetchType.LAZY)
+            @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL])
```
I dont think we want to cascaded the `result` and `exception` fields, since they dont always update. The blob might be fine to cascade.

lankydan reviewed (commented):
I'll have another run through this.

williamvigorr3 committed:
Fixed typos in Liquidbase script

Also corrected constraints and added missed fields in hibernate
checkpoint class and liquibase scripts.

williamvigorr3 committed:
Update CheckpointStorage to pass in serialization context.

This is cleaner than passing both the checkpoint and the
serialized checkpoint into the methods. Also fixed
CordaPersistanceServiceTests which I accidentally broke.

williamvigorr3 committed:
Fix detekt problem

williamvigorr3 committed:
Revert "Update CheckpointStorage to pass in serialization context."

This reverts commit b71e78f20274ab0f5b3cf3fda1451ae2bd7a6797.

williamvigorr3 committed:
Fix test broken by reverting commit

lankydan committed:
CORDA-3597 Update metadata join, timestamp columns and serialization

- Change the metadata join to the checkpoints table to use
  `invocation_id` instead of `flow_id`. There were issues joining
  between the tables because `flow_id` was not the primary key of the
  metadata table. Switching over to `invocation_id` has at least allowed
  us to bypass this issue. The information about the `invocation_id` is
  stored in the `Checkpoint` class which makes it simple to save at
  runtime.
- Some of timestamp columns were nullable when they should always be
  populated, the nullable flags have now been removed.
- Previously the whole checkpoint was being serialized and stored into
  the `checkpoints_blob.checkpoint` column. This meant duplicated saving
  as the `flow_state` was contained in this object. Only the
  `CheckpointState` property of `Checkpoint` is now being serialized and
  saved to this field. Furthermore, it now uses the default
  `STORAGE_CONTEXT` serialization (AMQP) instead of Kryo (which is only
  used for serializing the `flow_state` / flow stack).
- The checkpoint database performance metrics recording has been
  abstracted to its own class.

lankydan committed:
CORDA-3597 Make metadata join non optional

Remove the nullable declaration on the metadata field of
`DBFlowCheckpoint`

lankydan committed:
CORDA-3597 Rename `node_checkpoints_blobs` to `node_checkpoint_blobs`

lankydan committed:
CORDA-3597 Update some kdocs

roastario commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -2,245 +2,392 @@ package net.corda.node.services.persistence
-            @OneToOne(fetch = FetchType.LAZY)
-            @JoinColumn(name = "result_id", referencedColumnName = "id")
-            var result: DBFlowResult?,
+        @OneToOne(fetch = FetchType.LAZY, cascade = [CascadeType.ALL], optional = true)
```
By putting Cascade = ALL, we are forcing hibernate to update all the child tables when persisting this entity. I think this will have a pretty big performance impact. 

I would instead manually persist the child entities.

mnesbit reviewed (approved):
Basically makes sense to me
